lib/client.ts

```diff
@@ -88,35 +76,49 @@ export class Client {
     this.authorize(this.request(this.url(endpoint), { method: 'GET' }, reqConfig))
 
   public post = (endpoint: Endpoint, payload: any, reqConfig?: RequestConfig): Promise<any> =>
-    this.authorize(this.request(this.url(endpoint), { method: 'POST' , body: JSON.stringify(payload)}, reqConfig))
+    this.authorize(this.request(this.url(endpoint), { method: 'POST', body: JSON.stringify(payload) }, reqConfig))
 
   public patch = (endpoint: Endpoint, payload: any, reqConfig?: RequestConfig): Promise<any> =>
-    this.authorize(this.request(this.url(endpoint), { method: 'PATCH', body: JSON.stringify(payload)}, reqConfig))
+    this.authorize(this.request(this.url(endpoint), { method: 'PATCH', body: JSON.stringify(payload) }, reqConfig))
 
   public destroy = (endpoint: Endpoint, payload?: any, reqConfig?: RequestConfig): Promise<any> =>
-    this.authorize(this.request(this.url(endpoint), { method: 'DELETE', body: JSON.stringify(payload)}, reqConfig))
-
-  private url = (endpoint: Endpoint): string => ifElse(
-    is(String),
-    concat(this.host),
-    this.urlConfig
-  )(endpoint)
-
-  private urlConfig = ({ base, action, params = {}, qs }: EndpointConfig): string => compose(
-    when(
-      () => notNil(qs),
-      finalUrl => `${finalUrl}?${stringify(qs, { arrayFormat: 'brackets' })}`
-    ),
-    when(
-      () => notNil(action),
-      resourceUrl => `${resourceUrl}/${action}`
-    ),
-    when(
-      () => notNil(params.id),
-      collectionUrl => `${collectionUrl}/${params.id}`
-    ),
-    (hostname) => `${hostname}${S(base).template(params, '{', '}').s}`
-  )(this.host)
+    this.authorize(this.request(this.url(endpoint), { method: 'DELETE', body: JSON.stringify(payload) }, reqConfig))
+
+  private url = (endpoint: Endpoint): string =>
+    notNil(endpoint) && (endpoint.constructor === String || endpoint instanceof String) ?
+      `${this.host}${endpoint}` : // TODO: Do we want to allow this. Version will not be handled for this
+      this.urlConfig(endpoint as EndpointConfig);
+
+  private urlConfig = ({ base, action, params = {}, qs, apiVersion }: EndpointConfig): string => {
+    let url = `${this.host}/${this.version(apiVersion)}${replaceParams(base, params)}`;
+
+    if (notNil(params.id)) {
+      url = `${url}/${params.id}`;
+    }
+
+    if (notNil(action)) {
+      url = `${url}/${action}`;
+    }
+
+    if (notNil(qs)) {
+      url = `${url}?${stringify(qs, { arrayFormat: 'brackets' })}`;
+    }
+
+    return url;
+  };
+
+  private version = (apiVersion: string = 'v1.0'): string => {
```

Comment:
Do we need to process `''` (passed empty string from the client code) as default behaviour (`v1.0`) or we should throw an error in this case? (as now)